syntax = "proto3";

package orbax_model_manifest;

import "orbax/experimental/model/core/protos/type.proto";

message Manifest {
  map<string, TopLevelObject> objects = 1;
  optional GlobalSupplemental supplemental_info = 2;
  optional DeviceMesh device_mesh = 3;
  optional DeviceAssignmentByCoords device_assignment_by_coords = 4;

  // A list of paths frequently used in this manifest. Once a path is added to
  // this list, use sites can refer to it by its index here instead of spelling
  // out the full string.
  repeated string frequent_paths = 5;
}

message GlobalSupplemental {
  oneof case {
    // TODO(wangpeng): Consider removing the `single` case.
    UnstructuredData single = 1;
    GlobalSupplementalMap multiple = 2;
  }
}

message GlobalSupplementalMap {
  map<string, UnstructuredData> map = 1;
}

message DeviceMesh {
  message Axis {
    string name = 1;
    int32 size = 2;
  }

  repeated Axis axis = 1;
}

// Copied from /third_party/australis/google/ifrt/ifrt_australis.proto.
// Represents a mapping between device ids and coords. This allows consistently
// mapping between incompatible id spaces by aligning the coords.
message DeviceAssignmentByCoords {
  message Device {
    int64 id = 1;
    repeated int64 coords = 2;
    optional int64 core_on_chip = 3;
  }

  repeated Device devices = 1;
}

message TopLevelObject {
  oneof case {
    Function function = 1;
    Value value = 2;
  }
}

message Function {
  orbax_model_type.FunctionSignature signature = 1;
  FunctionBody body = 2;
  Visibility visibility = 3;
  optional string gradient_function_name = 4;
}

message Value {
  oneof case {
    ExternalValue external = 1;
    TupleValue tuple = 2;
    NamedTupleValue named_tuple = 3;

    // ...  list, dict, none
  }
}

message ExternalValue {
  optional orbax_model_type.Type type = 1;
  UnstructuredData data = 2;
}

message TupleValue {
  repeated string elements = 1;  // string is a name in Manifest.objects
}

message NamedTupleValue {
  message Pair {
    string name = 1;
    string value = 2;
  }

  repeated Pair elements = 1;
}

enum Visibility {
  PRIVATE = 0;
  PUBLIC = 1;
}

message FunctionBody {
  oneof case {
    StableHloFunctionBody stable_hlo_body = 1;
    UnstructuredData other = 2;  // e.g. TF graph
  }
}

message StableHloFunctionBody {
  UnstructuredData stable_hlo = 1;

  // some metadata
  uint32 calling_convention_version = 2;
  repeated string lowering_platforms = 3;
  repeated uint32 module_kept_var_idx = 4;
  optional UnstructuredData supplemental_info = 5;  // e.g. JAX-specific info
}

message UnstructuredData {
  oneof data {
    FileSystemLocation file_system_location = 1;
    string inlined_string = 2;
    bytes inlined_bytes = 3;
  }

  optional string mime_type = 4;  // e.g. “mlir_stablehlo”
  optional string version = 5;
}

message FileSystemLocation {
  oneof path {
    // an absolute or relative path (as string)
    string string_path = 1;

    // an index into `Manifest.frequent_paths`
    int32 frequent_path_idx = 2;
  }

  // Absence means 0 (i.e. start of the file).  Negative means counting from the
  // end, i.e. real_offset_in_byte = offset_in_byte + file_length_in_byte .
  optional int64 offset_in_byte = 3;

  // Absence means spanning to the end of the file.
  optional uint64 size_in_byte = 4;
}
